keps/sig-cli/5295-kyaml/README.md

```diff
@@ -0,0 +1,1228 @@
+
+Maps are rendered with curly-braces. Keys are sorted and are not quoted unless they are type-ambiguous (e.g. `no`), and values are rendered as per their type.
+
+<<[UNRESOLVED quoted strings]>>
+Should we treat strings for field names and map keys:
+As described above, with label keys quoted but field names not
+Allow unquoted label keys and more complex (but not ambiguous) values
+Always quote strings, including ALL map keys and field names
+<<[/UNRESOLVED>>
+
+Slices and arrays are rendered with square-braces. Elements are rendered as per this list.
+
+Pointers are rendered as `null` or dereferenced and rendered as per the pointee’s type.
+
+Types which implement the Go standard `json.Marshaler` interface (`MarshalJSON()`) are rendered by marshaling to JSON, unmarshaling into an `any`, and then rendering that `any` as KYAML.
+
+Types which implement the `yaml.Marshaler` interface (`MarshalYAML()`) are rendered as per their type.  `MarshalYAML()` is not a Go standard and is not used by KYAML.
+
+<<[UNRESOLVED yaml self-marshaling]>>
+Should we do something for YAML self-marshalling?  It is non-standard and we do not use it in Kubernetes.
+<<[/UNRESOLVED>>
+
+Interfaces are rendered as `null` or their true value type.
+
+Any other type (as per `reflect.Value.Kind()`) will produce an error.
+
+### Multi-line strings
+
+For better or worse, multi-line strings are a thing that people do in Kubernetes, and conventional YAML makes it pretty easy. KYAML has to choose how we want to handle it.  Consider these examples of conventional YAML.
+
+Example 1:
+```
+example_1: |
+    This
+    is a
+    multi-line string
+```
+
+Example 2:
+```
+example_2: |
+    this:
+      is:
+      - embedded
+      - yaml
+      where:
+        whitespace:
+          matters: true
+```
+
+#### Option 1. Treat it like a Go quoted-and-escaped string
+
+This always comes out as one line.
+
+```
+example_1: “This\nis a\nmulti-line string\n”
+example_2: "this:\n  is:\n  - embedded\n  - yaml\n  where:\n    whitespace:\n      matters: true\n"
+```
+
+Pro: Easy and well-defined
+Pro: Can be pasted into most C-like languages to get the original string.
+Con: Hard to read and write by humans when it gets long.
+
+#### Option 2. Use YAML flow-folding and escaped line-breaks
+
+This gets closer to the original without sacrificing too much of the goals of this KEP.
+
+```
+example_1: ”\
+    This\n\
+    is a\n\
+    multi-line string\
+    “
+example_2: "\
+    this:\n\
+    \  is:\n\
+    \  - embedded\n\
+    \  - yaml\n\
+    \  where:\n\
+    \    whitespace:\n\
+    \      matters: true\n\
+    "
+```
+
+Pro: Easier to read than option 1 and close to the original
+Pro: Closer to the original than option 3
+Con: Requires weird EOL handling (`\n\` meaning a newline, plus a line-continuation which  prevents the fold from creating a single leading space on the next line)
+Con: Subtle handling of leading whitespace (all dropped unless escaped)
```

Comment:
Worth noting: This is true of all multi-line yaml strings *except* option 1), even in the flow string styles we can't use.